projects/graph/src/graph.py

```diff
@@ -1,9 +1,83 @@
+from queue import Queue
 """
 Simple graph implementation
 """
 
 
 class Graph:
     """Represent a graph as a dictionary of vertices mapping labels to edges."""
+
     def __init__(self):
-        pass  # TODO
+        self.vertices = {}
+
+    def add_vertex(self, vertex):
+        self.vertices[vertex] = set()
+
+    def add_edge(self, vertex1, vertex2):
+        if vertex1 in self.vertices and vertex2 in self.vertices:
+            self.vertices[vertex1].add(vertex2)
+            self.vertices[vertex2].add(vertex1)
+        else:
+            undefined = []
+            if vertex1 not in self.vertices:
+                undefined.append(vertex1)
+            if vertex2 not in self.vertices:
+                undefined.append(vertex2)
+            undefined = ", ".join(undefined)
+            raise KeyError(f"Vertex not found: {undefined}")
+
+    def breadth_first_traversal(self, starting_node):
+        queue = Queue()
+        visited = set()
```

Comment:
I think it would make sense to use lists instead of sets in all of the places where you have `visited` because lists are ordered and sets are not.

Reply:
I have it as a set, because when we're checking if we visited a vertex or not it shouldn't matter if the set is in order. We're checking if the vertex is a member of visited to determine if we need to visit or not. Which I believe a set is faster than a list when checking for a value inside.

Reply:
Okay, thanks for letting me know. Good answer. 👍